[webhookeventreceiver] add option to include headers as attributes

tredman opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description

Hi! I'm looking to add a feature to the webhookeventreceiver to allow HTTP request headers to be passed through as log attributes. Our initial use case for this is for receiving [webhook events from Github/Github Enterprise](https://docs.github.com/en/enterprise-server@3.14/webhooks/webhook-events-and-payloads) - which stores important metadata in the webhook request headers that is absent from the payload.

This PR adds a new option, `convert_headers_to_attributes`, which will include any headers except the one specified in `required_header` in the payload as an attribute. 

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes

N/A - started with this PR

<!--Describe what testing was performed and which tests were added.-->
#### Testing

Added unit tests and I'm running this successfully in our environment using a local build.

<!--Describe the documentation added.-->
#### Documentation

README

<!--Please delete paragraphs that you did not use before submitting.-->

tredman committed:
[webhookeventreceiver] add option to include headers as attributes

tredman committed:
committing PR ID to changelog

shalper2 commented:
Hey this is really great! A couple of small things: First in `factory.go` the default config should be updated to include `false` for the new setting (since it is optional). Second, could you add two additional test cases for me though, one which covers the `normalizeHeader` function and another where `convert_headers_to_attributes` is set but no non-required headers are passed in? other than that I think this looks good!

tredman committed:
address PR feedback

- set ConvertHeadersToAttributes to false in `createDefaultConfig`
- Adds additional test cases for ReqToLog
- Adds tests for normalized header attribute key

tredman commented:
Thanks for the quick review and feedback! I think I've addressed it in my latest commit but let me know.

re: `normalizeHeader` - i renamed this to be a little more descriptive of what it was doing: generating an OTEL attribute key from a header.

shalper2 reviewed (approved):
lgtm! Thanks!

atoulme commented on receiver/webhookeventreceiver/req_to_log.go:
```diff
@@ -60,3 +67,21 @@ func appendMetadata(resourceLog plog.ResourceLogs, query url.Values) {
+		if k == textproto.CanonicalMIMEHeaderKey(config.RequiredHeader.Key) {
+			continue
+		}
+		scopeLog.Scope().Attributes().PutStr(headerAttributeKey(k), strings.Join(headers.Values(k), ";"))
```
You can put a slice there, no need to transform the headers into a string as a `;` separator will not work if any header value contains a `;`.

tredman replied:
ah I'm also missing a test case for multiple header values, will add one

atoulme commented on receiver/webhookeventreceiver/req_to_log.go:
```diff
@@ -60,3 +67,21 @@ func appendMetadata(resourceLog plog.ResourceLogs, query url.Values) {
+func headerAttributeKey(header string) string {
+	snakeCaseHeader := strings.ReplaceAll(header, "-", "_")
+	snakeCaseHeader = strings.ToLower(snakeCaseHeader)
+	return "header." + snakeCaseHeader
```
Any reason to transform to snake case? That can be done later in a processor

tredman replied:
Was (maybe poorly) trying to conform to https://opentelemetry.io/docs/specs/semconv/general/naming/

atoulme replied:
The issue is that this is input data, not a deterministic piece of data. You can have collision right now such as two headers with the names “content-length” and “Content Length”. It’s better to let processors perform data sanitization.

tredman replied:
Fair enough. For our use case it would indeed be easy enough to do this in a processor as you say, without baking it in here. What do you think about maintaining the `header.` namespace? I think it would still be useful to differentiate the attributes this way.

atoulme commented on receiver/webhookeventreceiver/config.go:
```diff
@@ -20,12 +20,13 @@ var (
+	Path                       string                   `mapstructure:"path"`                          // path for data collection. Default is /events
+	HealthPath                 string                   `mapstructure:"health_path"`                   // path for health check api. Default is /health_check
+	RequiredHeader             RequiredHeader           `mapstructure:"required_header"`               // optional setting to set a required header for all requests to have
+	ConvertHeadersToAttributes bool                     `mapstructure:"convert_headers_to_attributes"` // optional to convert all headers to attributes
```
Can we make this a regex instead? If empty, that setting doesn't apply. Otherwise, any header key mapping to the regular expression can be included. `.*` satisfies your use case.

tredman replied:
That does sound cleaner, would also help us filter out some unrelated headers that are just getting scooped up

tredman committed:
address PR feedback, refactor to use regex

tredman commented:
Updated PR to address feedback on multi-values and switch to using regex instead. I also added a comment to `headerAttributeKey` to hopefully clarify the attribute key naming

Switching to regex was a little more complex, since I wanted to ensure the regex was validated on startup and also pass the compiled regex down to reqToLog. I made a couple of changes:

- `reqToLog` is now a receiver function, receiving an `eventReciever` - this simplifies the function header but does mean it's a little more complex to test. I think this makes sense because reqToLog requires several values from the calling `eventReceiver` anyway
- `newLogsReceiver` now calls `Config.Validate()` - this was previously only referenced in tests, and there was some duplicated validation code. I added a couple of test cases for some previously untested max values. `Validate()` now also checks the regex.

tredman committed:
remove snake_case conversion for header attributes

mterhar commented:
I opened another PR against the same receiver. When this merges, I'll rebase. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38042

tredman commented:
:wave: Pushed a rebase after the last PR was merged. Is there any feedback here that still needs to be addressed?

songy23 reviewed (commented):
> Generated code is out of date, please run "make generate" and commit the changes in this PR.

also please fix lint http://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/13655296572/job/38267389141?pr=37815

tredman committed:
make generate, address lint errors

tredman commented:
ran `make generate` and fixed lint errors
